Adds logic to provide leniency for drag clicks

LemonInTheDark opened the pull request:
## About The Pull Request

Ok so like, in byond. if you click on something and release the click without moving off its sprite, that counts as a "click"

If you however move your mouse off its sprite, or it does so on its own, any further releases count as dragdrops.

We already have logic for catching drag -> mouse back over sprite. The code I'm adding here basically is just a bit nicer about how exact you need to be

I'm adding tweakable values for how much your mouse is allowed to move (This isn't perfect currently, since if the target itself moves we don't care, need to work on this) in a short time period. If you meet these cases, we instead call it a click.

They're globs for now cause I don't have a good feel for em, will tweak over time. I want to reduce dumb clickspam and click dragging moving mobs/drifting items by accident.

Oh also fixed a bug in ai behaviors that came from not knowing what get_active_hand did

Fixes a monkey behavior of using their held item

It turns out it never worked, and the only reason it didn't runtime was because of an if() check that got passed the result of an invalid proc call. this fixes that

## Why It's Good For The Game

Spam clicking and drag clicking are stupid. they should be less stupid

## Changelog
:cl:
add: Adds a grace period for hit detection, between clicking ON someone, and then moving your mouse off before releasing the click. It's short, hopefully it helps with stupid
/:cl:

LemonInTheDark committed:
Adds logic to provide leniency for drag clicks

Ok so like, in byond. if you click on something and release the click
without moving off its sprite, that counts as a "click"

If you however move your mouse off its sprite, or it does so on its own,
any further releases count as dragdrops.

We already have logic for catching drag -> mouse back over sprite.
The code I'm adding here basically is just a bit nicer about how exact
you need to be

I'm adding tweakable values for how much your mouse is allowed to move (This
isn't perfect currently, since if the target itself moves we don't care)
in a short time period. If you meet these cases, we instead call it a
click.

They're globs for now cause I don't have a good feel for em, will tweak
over time. I want to reduce dumb clickspam and click dragging moving
mobs/drifting items by accident.

san7890 commented on code/_onclick/drag_drop.dm:
```diff
@@ -10,14 +10,37 @@
+// These two are globs for the purpose of testing, convert to defines when you're done
+// Distance in pixels that we consider "acceptable" from the initial click to the release
+// Note: this does not account for the position of the object, just where it was when we first started the timer
+GLOBAL_VAR_INIT(lieniency_distance, 10)
+// Accepted time in seconds between the initial click and drag release
+// Go higher then this and we just don't care anymore
+GLOBAL_VAR_INIT(lieniency_time, 0.1 SECONDS)
```
```suggestion
GLOBAL_VAR_INIT(leniency_distance, 10)
// Accepted time in seconds between the initial click and drag release
// Go higher then this and we just don't care anymore
GLOBAL_VAR_INIT(leniency_time, 0.1 SECONDS)
```

LemonInTheDark replied:
they're temporary so no

san7890 commented on code/_onclick/drag_drop.dm:
```diff
@@ -10,14 +10,37 @@
+/client/proc/check_drag_proximity(atom/dragging, atom/over, params)
+	if(dragging == over)
+		return TRUE
+	if(world.time - drag_start > GLOB.lieniency_time) // Time's up bestie
```
```suggestion
	if(world.time - drag_start > GLOB.leniency_time) // Time's up bestie
```

san7890 commented on code/_onclick/drag_drop.dm:
```diff
@@ -10,14 +10,37 @@
+	var/list/new_offsets = screen_loc_to_offset(LAZYACCESS(modifiers, SCREEN_LOC), view)
+
+	var/distance = sqrt(((old_offsets[1] - new_offsets[1]) ** 2) + ((old_offsets[2] - new_offsets[2]) ** 2))
+	if(distance > GLOB.lieniency_distance)
```
```suggestion
	if(distance > GLOB.leniency_distance)
```

san7890 reviewed (commented):
spellcheck

LemonInTheDark commented:
~~I appreciate it but they're temporary vars, for testin~~

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into hit-detection

Mothblocks commented:
update?

LemonInTheDark commented:
Ok so like, it works, but it's kinda hard to test if it does what I want because it's not something people really notice in play.
I might uh, play a few rounds as a janitor or something and modify it between games until it feels right to me.  Not really sure how to approach this otherwise since the cases where it matters are sorta rare

Mothblocks commented:
Log when it happens and see idk

LemonInTheDark commented:
> Log when it happens and see idk

didn't even consider it, thanks bestie

LemonInTheDark committed:
Adds logging for hit detection lieniency, to see if it works or not

Also fixes a bug in monkey behaviors, ninja used the WRONG proc, was
trying to use an arm as a slot

LemonInTheDark committed:
Adds additional lieniency

If you click on a turf and drag to a non turf within the window
we assume you meant to click on the non turf, and just missed

SkeletalElite commented:
I pretty frequently accidentally kick/punch someone when trying to fireman carry or strip when this is merged, is a little annoying.

LemonInTheDark commented:
Should add a case for landing on a mob, since that's a lot of cases. I didn't think people dragged quickly enough for it to be a problem

LemonInTheDark committed:
Adds a case for dragging onto bucklable things/the eye of the client

LemonInTheDark committed:
Fixes a monkey behavior of using their held item

It turns out it never worked, and the only reason it didn't runtime
was because of an if() check that got passed the result of an invalid
proc call. this fixes that

LemonInTheDark committed:
fixes switching hands

optimumtact commented:
Finish this please

LemonInTheDark commented:
I think I'm happy with this, tho it can be tweaked in future. lemme undebug it

LemonInTheDark committed:
Removes debug variables, increases distance slightly

LemonInTheDark commented:
Finished
